src/charm.py

```diff
@@ -4,101 +4,424 @@
+                    "command": "/usr/local/bin/traefik",
+                },
             },
         }
-        # Add intial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
-        container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ActiveStatus()
 
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
+        current_pebble_layer = container.get_plan().to_dict()
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
+        if current_pebble_layer != updated_pebble_layer:
+            self.unit.status = MaintenanceStatus(f"updating the '{_TRAEFIK_SERVICE_NAME}' service")
 
-        Learn more about config at https://juju.is/docs/sdk/config
-        """
-        current = self.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
+            container.add_layer(_TRAEFIK_LAYER_NAME, updated_pebble_layer, combine=True)
 
-    def _on_fortune_action(self, event):
-        """Just an example to show how to receive actions.
+            try:
+                if is_restart := container.get_service(_TRAEFIK_SERVICE_NAME).is_running():
+                    self.unit.status = MaintenanceStatus(
+                        f"stopping the '{_TRAEFIK_SERVICE_NAME}' service to update the configurations"
+                    )
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle actions, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the actions.py file.
+                    container.stop(_TRAEFIK_SERVICE_NAME)
+            except Exception:
+                # We have not yet set up the pebble service, nevermind
+                logger.exception(
+                    "The following error occurred while stopping the '%s' service, "
+                    "maybe it has not been created yet?",
+                    _TRAEFIK_SERVICE_NAME,
+                    exc_info=True,
+                )
 
-        Learn more about actions at https://juju.is/docs/sdk/actions
-        """
-        fail = event.params["fail"]
-        if fail:
-            event.fail(fail)
+            self.unit.status = MaintenanceStatus(f"starting the '{_TRAEFIK_SERVICE_NAME}' service")
+
+            container.start(_TRAEFIK_SERVICE_NAME)
+
+            if is_restart:
+                logger.info(f"'{_TRAEFIK_SERVICE_NAME}' service restarted")
```

Comment:
Don't build logger strings :)

Reply:
Please ;-) DRY here trumps a negligible amount of performance.